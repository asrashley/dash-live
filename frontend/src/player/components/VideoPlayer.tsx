import { useRef } from "preact/hooks";
import { PlaybackIcon } from "./PlaybackIcon";
import { VideoElement, VideoElementProps } from "./VideoElement";

export type VideoPlayerProps = Omit<VideoElementProps, 'subtitlesElement'>;

export function VideoPlayer(props: VideoPlayerProps) {
  const subtitlesElement = useRef<HTMLDivElement>();

  return (
    <div id="vid-window">
      <PlaybackIcon active={props.activeIcon} />
      <VideoElement {...props}  />
    </div>
  );
}
