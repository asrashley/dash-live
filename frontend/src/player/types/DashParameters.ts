import type { OptionMapWithChildren } from "@dashlive/options";
import type { PeriodJson } from "./PeriodsJson";

export interface DashParameters {
    dash: {
        locationURL: string;
        mediaDuration: string;
        minBufferTime: string;
        mpd_id: string;
        now: string;
        periods: PeriodJson[];
        profiles: string[];
        publishTime: string;
        startNumber: number;
        suggestedPresentationDelay: number;
        timeSource: null;
        title: string;
    },
    options: OptionMapWithChildren;
    url: string;
}
