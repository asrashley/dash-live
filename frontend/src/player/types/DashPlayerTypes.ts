export type DashPlayerTypes = 'native' | 'shaka' | 'dashjs';