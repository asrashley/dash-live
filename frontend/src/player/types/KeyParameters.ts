export type KeyParameters = {
    alg: string;
    kid: string; // hex encoded
    key: string; // hex encoded
    guidKid: string;
    b64Key: string; // base64 encoded
    computed: boolean;
};
