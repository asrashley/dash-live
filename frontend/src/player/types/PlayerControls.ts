export interface PlayerControls {
    isPaused(): boolean;
    pause(): void;
    play(): void;
    skip(seconds: number): void;
    stop(): void;
}