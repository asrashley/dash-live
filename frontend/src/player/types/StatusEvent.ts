export type StatusEvent = {
    id: number;
    position: number;
    timecode: string;
    event: string;
    text: string;
};