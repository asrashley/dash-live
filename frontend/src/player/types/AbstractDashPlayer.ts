import type { OptionMapWithChildren } from "@dashlive/options";
import { KeyParameters } from "./KeyParameters";

export interface DashPlayerProps {
    version?: string;
    autoplay?: boolean;
    videoElement: HTMLVideoElement;
    subtitlesElement?: HTMLDivElement;
    logEvent(eventName: string, text: string): void;
}
export abstract class AbstractDashPlayer {

    // eslint-disable-next-line no-useless-constructor
    constructor(protected props: DashPlayerProps) {
        /* no op */
    }

    abstract initialize(mpd: string, options: OptionMapWithChildren, keys: Map<string, KeyParameters>): Promise<void>;

    abstract destroy(): void;

    protected onCanPlayEvent = async () => {
        const { videoElement } = this.props;
        try {
            await videoElement.play();
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        } catch(err) {
            videoElement.muted = true;
            videoElement.play();
        }
        videoElement.removeEventListener('canplay', this.onCanPlayEvent);
    };

}