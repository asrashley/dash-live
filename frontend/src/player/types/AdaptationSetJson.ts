import { DrmContextMap } from "./DrmContextMap";
import { KeyParameters } from "./KeyParameters";

export type AdaptationSetJson = {
    content_type: string;
    default_kid: string | null;
    drm?: DrmContextMap;
    fileSuffix: string;
    id: number;
    initURL: string;
    keys?: { [kid: string]: KeyParameters };
    lang: null | string;
    mediaURL: string;
    mimeType: string;
    mode: string;
    par: string;
    segmentAlignment: boolean;
    startWithSAP: number;
    timescale: number;
};
