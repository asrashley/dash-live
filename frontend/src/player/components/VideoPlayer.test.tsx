import { describe, expect, test } from "vitest";
import { signal } from "@preact/signals";

import { renderWithProviders } from "../../test/renderWithProviders";
import { VideoPlayer } from "./VideoPlayer";
import { DashParameters } from "../types/DashParameters";
import { KeyParameters } from "../types/KeyParameters";
import { PlayerControls } from "../types/PlayerControls";
import { StatusEvent } from "../types/StatusEvent";

describe("VideoPlayer component", () => {
  const params: DashParameters = {
    dash: {
      locationURL: "https://unit.test.local/test.mpd",
      mediaDuration: "PT30S",
      minBufferTime: "PT4S",
      mpd_id: "mpd-id",
      now: "2025-04-05T01:02:03Z",
      periods: [],
      profiles: [],
      publishTime: "2025-01-01T00:00:00Z",
      startNumber: 1,
      suggestedPresentationDelay: 0,
      timeSource: null,
      title: "VideoElement test",
    },
    options: {},
    url: "https://unit.test.local/test.mpd",
  };
  const dashParams = signal<DashParameters>(params);
  const keys = signal<Map<string, KeyParameters>>(new Map());
  const currentTime = signal<number>(0);
  const controls = signal<PlayerControls | undefined>();
  const events = signal<StatusEvent[]>([]);

  test("matches snapshot", () => {
    const { asFragment } = renderWithProviders(<VideoPlayer
        mpd={params.url}
        playerName="native"
        dashParams={dashParams}
        keys={keys}
        currentTime={currentTime}
        controls={controls}
        events={events}
    />);
    expect(asFragment()).toMatchSnapshot();
  });
});
