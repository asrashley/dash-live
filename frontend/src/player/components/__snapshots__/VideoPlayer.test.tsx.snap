// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`VideoPlayer component > matches snapshot 1`] = `
<DocumentFragment>
  <div
    id="vid-window"
  >
    <video
      src="https://unit.test.local/test.mpd"
    />
  </div>
</DocumentFragment>
`;
